Mage/src/main/java/mage/watchers/Watcher.java

```diff
@@ -75,18 +77,61 @@ public void reset() {
         condition = false;
     }
 
-    protected String getBasicKey(){
+    protected String getBasicKey() {
         return getClass().getSimpleName();
     }
 
     public abstract void watch(GameEvent event, Game game);
 
-    public <T extends Watcher> T copy(){
+    public <T extends Watcher> T copy() {
         try {
-            Constructor<? extends Watcher> constructor = this.getClass().getDeclaredConstructor(getClass());
+            List<?> constructors = Arrays.asList(this.getClass().getConstructors());
+            if (constructors.size() > 1) {
+                logger.error(getClass().getSimpleName() + " has multiple constructors");
+                return null;
+            }
+
+            Constructor<? extends Watcher> constructor = (Constructor<? extends Watcher>) constructors.get(0);
+            if (constructor.getParameterCount() > 0) {
+                logger.error(getClass().getSimpleName() + " constructor has arguments, should be 0 and inject the parameters by setters");
+                return null;
+            }
             constructor.setAccessible(true);
-            return (T) constructor.newInstance(this);
-        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            T watcher = (T) constructor.newInstance();
+            List<Field> allFields = new ArrayList<>();
+            allFields.addAll(Arrays.asList(getClass().getDeclaredFields()));
+            allFields.addAll(Arrays.asList(getClass().getSuperclass().getDeclaredFields()));
+            for (Field field : allFields) {
+                field.setAccessible(true);
+                if (field.getType().isPrimitive()) {
+                    field.set(watcher, field.get(this));
+                }
+                else if(field.getType() == Step.class){
+                    field.set(watcher, field.get(this));
+                }
+                else if (field.getType() == Mana.class) {
+                    field.set(watcher, field.get(this));
+                } else if (field.getType() == UUID.class) {
+                    field.set(watcher, field.get(this));
+                } else if (field.getType().isEnum()) {
+                    field.set(watcher, field.get(this));
+                } else if (field.getType() == Set.class) {
+                    ((Set) field.get(watcher)).clear();
+                    ((Set) field.get(watcher)).addAll((Set) field.get(this));
+                } else if (field.getType() == Map.class) {
+                    ((Map) field.get(watcher)).clear();
+                    ((Map) field.get(watcher)).putAll((Map) field.get(this));
+                } else if (field.getType() == List.class) {
+                    ((List) field.get(watcher)).clear();
```

Comment:
If for some reason you wanted to support immutable lists, you could just do `new ArrayList((Collection) field.get(this));`

Reply:
At runtime I do not have access to the implementation of the List. List is an interface, ArrayList is an implementation. So that is why I clear the current list / collection, and re-add all items. 
For example, we use in xmage different kind of Maps : EnumMap, HashMap, etc. It would be unsafe to use new HashMap<>() in this copy method, when in the original implementation the Map was an EnumMap.  So in short: use interfaces, not implementations!

Reply:
I see why you might not want to change the implementation class (performance, side-effects, etc).  The interface is satisfied by any implementation (that implements it), so it shouldn't matter what the concrete implementation is.

If the watcher has something like this:
```
Set<Things> things = EnumSet.of(Thing.ONE, Thing.TWO);
```
then it should be fine (compile and execute correctly) to do this:
```
things = new HashSet<Things>(things);
```